# Calculadora de Bayes

Este notebook acompana el tema de probabilidad y sistemas expertos probabilisticos. Calcula una probabilidad posterior usando el teorema de Bayes.

## Problema

En una region, la probabilidad de que un adulto mayor de 40 anos tenga cancer es `P(C) = 0.05`. Un diagnostico detecta correctamente la enfermedad con `P(D|C) = 0.78` y produce un falso positivo con `P(D|no C) = 0.06`. Se desea calcular `P(C|D)`: la probabilidad de tener cancer dado un diagnostico positivo.

## Formula

```text
P(A|B) = P(B|A) * P(A) / P(B)
P(B) = P(B|A) * P(A) + P(B|no A) * P(no A)
```

![Formula de Bayes](../assets/imagenes/bayes_formula.png)

In [1]:
def calcular_bayes(prob_a, prob_b_dado_a, prob_b_dado_no_a):
    """Calcula P(A|B) con el teorema de Bayes.

    Args:
        prob_a: probabilidad previa P(A).
        prob_b_dado_a: sensibilidad o P(B|A).
        prob_b_dado_no_a: falso positivo o P(B|no A).

    Returns:
        Probabilidad posterior P(A|B).
    """
    valores = {
        "P(A)": prob_a,
        "P(B|A)": prob_b_dado_a,
        "P(B|no A)": prob_b_dado_no_a,
    }
    for nombre, valor in valores.items():
        if not 0 <= valor <= 1:
            raise ValueError(f"{nombre} debe estar entre 0 y 1")

    prob_no_a = 1 - prob_a
    prob_b = (prob_b_dado_a * prob_a) + (prob_b_dado_no_a * prob_no_a)
    if prob_b == 0:
        raise ValueError("P(B) no puede ser cero")
    return (prob_b_dado_a * prob_a) / prob_b

In [2]:
prob_cancer = 0.05
prob_diagnostico_si_cancer = 0.78
prob_diagnostico_si_no_cancer = 0.06

resultado = calcular_bayes(
    prob_cancer,
    prob_diagnostico_si_cancer,
    prob_diagnostico_si_no_cancer,
)

print(f"P(C|D) = {resultado:.5f} = {resultado:.2%}")

P(C|D) = 0.40625 = 40.63%


In [3]:
# Prueba rapida para validar que el calculo coincide con el ejercicio.
assert round(resultado, 5) == 0.40625

## Interpretacion

Aunque el diagnostico acierta con frecuencia cuando la persona tiene cancer, la enfermedad tiene una probabilidad previa baja. Por eso, despues de un resultado positivo, la probabilidad posterior calculada es aproximadamente `40.63%`, no `78%`. Este tipo de razonamiento es importante en sistemas expertos probabilisticos porque combina evidencia nueva con conocimiento previo.